core/src/server/host_filtering.rs

```diff
@@ -0,0 +1,243 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! HTTP Host Header validation.
+
+use crate::Error;
+use http::uri::{InvalidUri, Uri};
+use route_recognizer::Router;
+use std::str::FromStr;
+
+/// Port pattern
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+pub enum Port {
+	/// No port specified (default port)
+	Default,
+	/// Port specified as a wildcard pattern (*).
+	Any,
+	/// Fixed numeric port
+	Fixed(u16),
+}
+
+impl From<u16> for Port {
+	fn from(port: u16) -> Port {
+		Port::Fixed(port)
+	}
+}
+
+/// Represent the http URI scheme that is returned by the HTTP host header
+///
+/// <http-URI = "http:" "//" authority path-abempty [ "?" query ][ "#" fragment ]>
+///
+/// Further information can be found: https://www.rfc-editor.org/rfc/rfc7230#section-2.7.1
+#[derive(Clone, Hash, PartialEq, Eq, Debug)]
+pub struct Authority {
+	hostname: String,
+	port: Port,
+}
+
+impl FromStr for Authority {
+	type Err = String;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		let uri: Uri = s.parse().map_err(|e: InvalidUri| e.to_string())?;
+		let authority = uri.authority().ok_or_else(|| "HTTP Host must contain authority".to_owned())?;
+		let hostname = authority.host();
+		let maybe_port = &authority.as_str()[hostname.len()..];
```

Comment:
Ah; can't do `authority.port()` because port may bve eg `*` I guess!

Reply:
yes and also `http::Uri::port` won't tell us whether it has a default port or not as well, so we need to do some manual stuff here.

so a little bit hacky but this was cleanest I could come up with